tests/test_nhg.py

```diff
@@ -138,3 +139,194 @@ def test_route_nhg(self, dvs, testlog):
+            return fvs
+
+        MAX_ECMP_COUNT = 512
+        MAX_PORT_COUNT = 10
+        IP_INTEGER_BASE = int(ipaddress.IPv4Address(unicode("2.2.2.0")))
+
+        config_db = swsscommon.DBConnector(swsscommon.CONFIG_DB, dvs.redis_sock, 0)
+        intf_tbl = swsscommon.Table(config_db, "INTERFACE")
+        fvs = swsscommon.FieldValuePairs([("NULL", "NULL")])
+
+        for i in range(MAX_PORT_COUNT):
+            intf_tbl.set(port_name(i), fvs)
+            intf_tbl.set("{}|{}".format(port_name(i), port_ipprefix(i)), fvs)
+            dvs.runcmd("config interface startup " + port_name(i))
+            dvs.runcmd("arp -s {} {}".format(peer_ip(i), port_mac(i)))
+            assert dvs.servers[i].runcmd("ip link set down dev eth0") == 0
+            assert dvs.servers[i].runcmd("ip link set up dev eth0") == 0
+
+        db = swsscommon.DBConnector(0, dvs.redis_sock, 0)
+        ps = swsscommon.ProducerStateTable(db, "ROUTE_TABLE")
+
+        # Add first batch of routes with unique nexthop groups in AppDB
+        route_count = 0
+        r = 0
+        while route_count < MAX_ECMP_COUNT:
+            r += 1
+            fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+            binary = fmt.format(r)
+            # We need at least 2 ports for a nexthop group
+            if binary.count('1') <= 1:
+                continue
+            fvs = gen_nhg_fvs(binary)
+            route_ipprefix = gen_ipprefix(route_count)
+            ps.set(route_ipprefix, fvs)
+            route_count += 1
+
+        time.sleep(1)
+
+        adb = swsscommon.DBConnector(1, dvs.redis_sock, 0)
+        rtbl = swsscommon.Table(adb, "ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
+        nhgtbl = swsscommon.Table(adb, "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP")
+
+        # check ASIC DB on the count of nexthop groups used
+        nhg_count = len(nhgtbl.getKeys())
+        assert nhg_count == MAX_ECMP_COUNT
+
+        # Add second batch of routes with unique nexthop groups in AppDB
+        # Add more routes with new nexthop group in AppDBdd
+        route_ipprefix = gen_ipprefix(route_count)
+        base_ipprefix = route_ipprefix
+        base = route_count
+        route_count = 0
+        while route_count < 10:
+            r += 1
+            fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+            binary = fmt.format(r)
+            # We need at least 2 ports for a nexthop group
+            if binary.count('1') <= 1:
+                continue
+            fvs = gen_nhg_fvs(binary)
+            route_ipprefix = gen_ipprefix(base + route_count)
+            ps.set(route_ipprefix, fvs)
+            route_count += 1
+        last_ipprefix = route_ipprefix
+
+        time.sleep(1)
+
+        # Check ASIC DB on the count of nexthop groups used, and it should not increase
+        nhg_count = len(nhgtbl.getKeys())
+        assert nhg_count == MAX_ECMP_COUNT
+
+        # Check the route points to next hop group
+        k = asic_route_exists("2.2.2.0/24")
```

Comment:
@daall , is this the right api to use to do the assert?

Reply:
Nope, @qiluo-msft you'll want to take a look at the methods in `tests/dvs_database.py` and use those to get the keys/entries you need. You can fetch the asic DB/other DBs that you need with `get_asic_db()` and similar methods in the DockerVirtualSwitch in `conftest.py`. Let me know if you have any questions!

Reply:
This case should be ok, I already used `wait_for_n_keys` above so everything is available at this time.

---
In reply to: [419621718](https://github.com/Azure/sonic-swss/pull/1238#discussion_r419621718) [](ancestors = 419621718)

Reply:
I follow the new convention on above code, no need to change here since everything is available this time.

---
In reply to: [419608002](https://github.com/Azure/sonic-swss/pull/1238#discussion_r419608002) [](ancestors = 419608002)

Reply:
This should work! I would prefer if we stuck to the new interface (e.g. using `get` and `set` on the dvs db object) instead of using a mix of the new interface and the swsscommon interface, but I don't think any additional waits are needed.

Reply:
I did not use any db operation here.

---
In reply to: [420330242](https://github.com/Azure/sonic-swss/pull/1238#discussion_r420330242) [](ancestors = 420330242)